File: 卡码/代码随想录/Python/图论/dfs/LeetCode_0695.py
#695. 岛屿的最大面积
from typing import *

class Solution:
	def __init__(self):
		self.dir:List[tuple[int,int]]=[(0,1),(1,0),(0,-1),(-1,0)]
		self.cnt:int=1
	def dfs(self,grid:List[List[int]],x:int,y:int):
		grid[x][y]=0
		for i in range(4):
			nextX:int=x+self.dir[i][0]
			nextY:int=y+self.dir[i][1]
			if nextX<0 or nextX>len(grid)-1 or nextY<0 or nextY>len(grid[0])-1:
				continue
			if grid[nextX][nextY]==1:
				self.cnt+=1
				self.dfs(grid,nextX,nextY)
	def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
		n,m=len(grid),len(grid[0])
		res:int=0
		for i in range(n):
			for j in range(m):
				if grid[i][j]==1:
					self.cnt=1
					self.dfs(grid,i,j)
					res=max(res,self.cnt)
		return res

File: 卡码/代码随想录/Go/图论/dfs/LeetCode_0695.go
// 695. 岛屿的最大面积
package main

func max(a, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}

func maxAreaOfIsland(grid [][]int) int {
	n, m := len(grid), len(grid[0])
	cnt := 1
	res := 0
	dir := [2][4]int{[4]int{0, 1, 0, -1}, [4]int{1, 0, -1, 0}}
	var dfs func(x, y int)
	dfs = func(x, y int) {
		grid[x][y] = 0
		for i := 0; i < 4; i++ {
			nextX := x + dir[0][i]
			nextY := y + dir[1][i]
			if nextX < 0 || nextX > n-1 || nextY < 0 || nextY > m-1 {
				continue
			}
			if grid[nextX][nextY] == 1 {
				cnt++
				dfs(nextX, nextY)
			}
		}
	}
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if grid[i][j] == 1 {
				cnt = 1
				dfs(i, j)
				res = max(res, cnt)
			}
		}
	}
	return res
}
